File: LeetCode/[M] - 189. Rotate Array/solution.py
from typing import List


class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        lenNums = len(nums)
        i, j = lenNums - (k % lenNums), len(nums) - 1
        while i < j:
            nums[i], nums[j] = nums[j], nums[i]
            i += 1
            j -= 1
        i, j = 0, lenNums - (k % lenNums) - 1
        while i < j:
            nums[i], nums[j] = nums[j], nums[i]
            i += 1
            j -= 1
        i, j = 0, lenNums - 1
        while i < j:
            nums[i], nums[j] = nums[j], nums[i]
            i += 1
            j -= 1

File: LeetCode/[M] - 189. Rotate Array/solution.go
package main

func rotate(nums []int, k int) {
	reverse(nums[len(nums)-(k%len(nums)):])
	reverse(nums[:len(nums)-(k%len(nums))])
	reverse(nums)
}

func reverse(nums []int) {
	for i, j := 0, len(nums)-1; i < j; i, j = i+1, j-1 {
		nums[i], nums[j] = nums[j], nums[i]
	}
}
